### DISCUSSION TITLE
multigrid

### DISCUSSION THREAD

**Discussion by LionelGelebart** [xcompact3d/2decomp-fft#298] [General]
Is the new version of 2decomp able to account for 2 (at least) grids of different sizes in the same program (that can perform fft on each grid).
I did modifications in the old version of 2decomp to do this but now my code as diverged from yours...

**Comment by mathrack** [comment]
Hello,

Long story short : the FFT engine supports a single resolution. No multigrid option is available.

Long story long : I was working on in-place real-to-complex and in-place complex-to-real FFT. See https://github.com/2decomp-fft/2decomp-fft/pull/131 and https://github.com/2decomp-fft/2decomp-fft/pull/144. My plan was to merge this potential performance improvement and then refactor the FFT engines to allow multigrid.

However, as you can see, the in-place r2c / c2r has never been merged and the refactoring of the FFT engines has not progressed.

As a side remark, I am currently trying to refactor the IO engines. I hope my work on IO will be more successful than my work on FFT. However the code is relatively large and refactoring it is a bit difficult.

**Reply by LionelGelebart** [reply]
Hello, thank you for the stories!
By the way, I have contacted Paul Bartholomew directly by mail to have informations on the 'reborn' of 2decomp.  I also proposed him to have a look at the way I have modified the old version of 2decomp to account for multigrids. If interested, I can make you (or anyother 2decomp developer), member of the amitex project on our gitlab. Just tell me.

**Reply by mathrack** [reply]
Sounds interesting. I will contact you by email.

**Reply by LionelGelebart** [reply]
OK, I have made an extraction of the modified 2decomp library here : https://github.com/LionelGelebart/2decomp15_mg
The idea is that I have converted variables to pointers and I move frm simulation to another with pointer associations.
I'm pretty sure it can not be introduced 'as is' in 2decomp but it can be regarded as a starting point...
Feel free to comment

**Reply by mathrack** [reply]
I think we could merge something close to what is available in the file `fft_fftw3_f03.f90` in your project. The main advantage is that the modification does not require huge modifications in the FFT subroutines.

**Comment by rfj82982** [comment]
Thanks for the discussion. I'll have a look at it as well and see how we can merge this into the dev branch. 
About the issue with

**Reply by LionelGelebart** [reply]
OK, I have made a recent modification that I think simplifies the use of multigrids (see modif in examples/test_fft_r2c_mg).
I stop and let you read and tell me your opinion about it.
Bye and have a nice we!

**Comment by mathrack** [comment]
I have started to sketch the integration of the multigrid capability in the fftw3_f03 backend. See https://github.com/2decomp-fft/2decomp-fft/pull/302
